<a href="https://colab.research.google.com/github/AsalBotirova/ai-ml-course/blob/main/module-02-programming-and-math/assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Task 1: Import and Load the Data**

What to do: Upload the Superstore CSV to Colab and load it into a Pandas DataFrame.

In [ ]:
import pandas as pd

df = pd.read_csv("Sample - Superstore.csv", encoding='latin1')
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


**Task 2: Inspect the Structure**

What to do: Examine the dataset's shape, data types, missing values, and summary statistics.

In [ ]:
print("Shape: ", df.shape) #rows and columns
df.info() #Data types and missing values
df.describe() #Statistical summary

Shape:  (9994, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   fl

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


The dataset contains 9994 rows and 21 columns with no missing values across all columns. The data includes 15 object(text) columns, 3 float64 columns - notably, Order Date and Ship Date are stored as text(object) and will need to be converted to datetime format. Sales ranges from $0.44 to $22638, and Profit ranges from -$6,599 to $8,399, indicating high variability and the presence of loss-making orders.

**Task 3: Handle Missing Values**

What to do: Fill missing values in numerical columns with the median and categorical columns with the mode.


In [ ]:
from numpy import number
# Step 1: Check which columns have missing values
print("Missing values before cleaning:")
print(df.isnull().sum())

# Step 2: Fill missing values (good practice even if none exist now)
# Numerical columns → fill with median
numerical_cols = df.select_dtypes(include = 'number').columns
df[numerical_cols] = df[numerical_cols].fillna(df[numerical_cols].median())

# Categorical columns → fill with mode
categorical_cols = df.select_dtypes(include = "object").columns
for col in categorical_cols:
  df[col] = df[col].fillna(df[col].mode()[0])

# Step 3: Verify no missing values remain
print("\nMissing values after cleaning:")
print(df.isnull().sum())


Missing values before cleaning:
Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

Missing values after cleaning:
Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64


The dataset had no missing values, so no imputation was necessary. However, the cleaning code was applied as a precaution — numerical columns would be filled with the median to handle outlier robustness, and categorical columns with the most frequent value (mode). The final check confirms zero null values across all 21 columns.

**Task 4: Remove Duplicates**

What to do: Check for and remove duplicate rows.


In [ ]:
# Step 1: Count duplicate rows
print("Number of duplicate rows:", df.duplicated().sum)

# Step 2: Remove duplicates
df = df.drop_duplicates()

# Step 3: Report final shape
print("Rows removed:", 9994 - df.shape[0])
print("Final shape:", df.shape)

Number of duplicate rows: <bound method Series.sum of 0       False
1       False
2       False
3       False
4       False
        ...  
9989    False
9990    False
9991    False
9992    False
9993    False
Length: 9994, dtype: bool>
Rows removed: 0
Final shape: (9994, 21)


The dataset was checked for identical duplicate rows. After removing duplicates, the final shape is reported — if the row count matches 9,994, no duplicates existed. Data integrity is now ensured with all remaining rows being unique.

**Task 5: Convert Data Types**

What to do: Convert Order Date and Ship Date columns to the proper datetime format.


In [ ]:
# Step 1: Convert Order Date and Ship Date to datetime
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

# Step 2: Confirm the conversion
print(df[['Order Date', 'Ship Date']].dtypes)

# Step 3: Spot-check a few rows
df[['Order Date', 'Ship Date']].head(5)

Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object


,Order Date,Ship Date
0,2016-11-08,2016-11-11
1,2016-11-08,2016-11-11
2,2016-06-12,2016-06-16
3,2015-10-11,2015-10-18
4,2015-10-11,2015-10-18


The Order Date and Ship Date columns were originally stored as text (object type) and have been successfully converted to datetime64 format. This allows for proper date-based operations such as sorting, filtering, and calculating time differences. A spot-check of the first 5 rows confirms the dates are formatted correctly.

**Task 6: Create Derived Features**

What to do: Create three customer-level summary metrics: total spending, order frequency, and average order value.

In [ ]:
total_spending = df.groupby('Customer ID')['Sales'].sum().rename('Total Spending')

order_frequency = df.groupby('Customer ID')['Order ID'].nunique().rename('Order Frequency')

avg_order_value = df.groupby('Customer ID')['Sales'].mean().rename('Average Order Value')



customer_summary = pd.concat([total_spending, order_frequency, avg_order_value], axis=1,).reset_index()
customer_summary.head()


,Customer ID,Total Spending,Order Frequency,Average Order Value
0,AA-10315,5563.560,5,505.778182
1,AA-10375,1056.390,9,70.426000
2,AA-10480,1790.512,4,149.209333
3,AA-10645,5086.935,6,282.607500
4,AB-10015,886.156,3,147.692667


In [ ]:
# Save cleaned main dataset
df.to_csv('superstore_cleaned.csv', index = False)

# Save customer summary table
customer_summary.to_csv('customer_summary.csv', index=False)

print("Files saved successfully!")

Files saved successfully!
